File: ts/sdk/src/checkpoint.ts
/**
 * c2sp.org/tlog-checkpoint format for MTA-QR.
 *
 * Checkpoint body (the bytes signed by the issuer and witnesses):
 *   <origin>\n
 *   <tree_size decimal>\n
 *   <root_hash base64std_padded>\n
 *
 * Witnesses always use Ed25519 per c2sp.org/tlog-cosignature, regardless
 * of the issuer's sig_alg.
 */

import {
  createHash, randomBytes,
  sign as nodeSign, verify as nodeVerify,
  createPrivateKey, createPublicKey, KeyObject,
} from "crypto";

const ED25519_PKCS8_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
const ED25519_SPKI_PREFIX  = Buffer.from("302a300506032b6570032100", "hex");

function seedToPrivKey(seed: Uint8Array): KeyObject {
  const der = Buffer.concat([ED25519_PKCS8_PREFIX, Buffer.from(seed)]);
  return createPrivateKey({ key: der, format: "der", type: "pkcs8" });
}

function rawPubToKeyObject(pubBytes: Uint8Array): KeyObject {
  const der = Buffer.concat([ED25519_SPKI_PREFIX, Buffer.from(pubBytes)]);
  return createPublicKey({ key: der, format: "der", type: "spki" });
}

export function pubKeyFromSeed(seed: Uint8Array): Uint8Array {
  const priv = seedToPrivKey(seed);
  const pub  = createPublicKey(priv);
  return new Uint8Array((pub.export({ format: "der", type: "spki" }) as Buffer).slice(-32));
}

export function generateSeed(): Uint8Array {
  return new Uint8Array(randomBytes(32));
}

/** Build the checkpoint body bytes that are signed. */
export function checkpointBody(
  origin: string,
  treeSize: bigint | number,
  rootHash: Uint8Array,
): Uint8Array {
  const rootHashB64 = Buffer.from(rootHash).toString("base64");
  return new TextEncoder().encode(`${origin}\n${treeSize}\n${rootHashB64}\n`);
}

/** Sign a checkpoint body. Returns 64-byte Ed25519 signature. */
export function signCheckpointBody(body: Uint8Array, privKeySeed: Uint8Array): Uint8Array {
  return new Uint8Array(nodeSign(null, Buffer.from(body), seedToPrivKey(privKeySeed)));
}

/** Verify an Ed25519 checkpoint signature. */
export function verifyCheckpointSig(
  body: Uint8Array,
  sig: Uint8Array,
  pubKeyBytes: Uint8Array,
): boolean {
  try {
    return nodeVerify(null, Buffer.from(body), rawPubToKeyObject(pubKeyBytes), Buffer.from(sig));
  } catch { return false; }
}

/**
 * Build the cosignature/v1 signed message:
 *   cosignature/v1\n
 *   time <unix_timestamp_decimal>\n
 *   <checkpoint body>
 */
export function cosignatureMessage(body: Uint8Array, timestamp: bigint): Uint8Array {
  const header = new TextEncoder().encode(`cosignature/v1\ntime ${timestamp}\n`);
  const msg = new Uint8Array(header.length + body.length);
  msg.set(header);
  msg.set(body, header.length);
  return msg;
}

export function signCosignature(
  body: Uint8Array,
  timestamp: bigint,
  privKeySeed: Uint8Array,
): Uint8Array {
  return new Uint8Array(
    nodeSign(null, Buffer.from(cosignatureMessage(body, timestamp)), seedToPrivKey(privKeySeed))
  );
}

export function verifyCosignature(
  body: Uint8Array,
  timestamp: bigint,
  sig: Uint8Array,
  pubKeyBytes: Uint8Array,
): boolean {
  try {
    return nodeVerify(
      null,
      Buffer.from(cosignatureMessage(body, timestamp)),
      rawPubToKeyObject(pubKeyBytes),
      Buffer.from(sig),
    );
  } catch { return false; }
}

/**
 * Derive the 4-byte key ID per c2sp.org/signed-note:
 *   key_id = SHA-256(name || 0x0A || 0x01 || raw_pubkey)[0:4]
 * where 0x0A is newline and 0x01 is the Ed25519 signature type identifier byte.
 */
export function noteKeyId(name: string, pubKey: Uint8Array): Uint8Array {
  const h = createHash("sha256");
  h.update(Buffer.from(name, "utf8"));
  h.update(Buffer.from([0x0a, 0x01]));
  h.update(Buffer.from(pubKey));
  return new Uint8Array(h.digest().subarray(0, 4));
}

export { noteKeyId as witnessKeyId };

/** Compute origin_id: first 8 bytes of SHA-256(origin) as big-endian uint64. */
export function computeOriginId(origin: string): bigint {
  const h = createHash("sha256").update(origin).digest();
  let id = BigInt(0);
  for (let i = 0; i < 8; i++) id = (id << BigInt(8)) | BigInt(h[i]);
  return id;
}

export function parseCheckpointBody(body: Uint8Array): {
  origin: string;
  treeSize: bigint;
  rootHash: Uint8Array;
} {
  const text = new TextDecoder().decode(body);
  if (!text.endsWith("\n")) throw new Error("checkpoint body must end with \\n");
  const lines = text.slice(0, -1).split("\n");
  // Per c2sp.org/tlog-checkpoint: three mandatory lines plus optional extension lines.
  if (lines.length < 3) throw new Error(`checkpoint body must have at least 3 lines, got ${lines.length}`);
  const [origin, treeSizeStr, rootHashB64] = lines;
  const treeSize = BigInt(treeSizeStr);
  const rootHash = new Uint8Array(Buffer.from(rootHashB64, "base64"));
  if (rootHash.length !== 32) throw new Error(`root_hash must be 32 bytes`);
  return { origin, treeSize, rootHash };
}

File: go/shared/checkpoint/checkpoint.go
// Package checkpoint implements the c2sp.org/tlog-checkpoint format for MTA-QR.
//
// A checkpoint body is exactly three newline-terminated lines:
//
//	<origin>\n
//	<tree_size decimal>\n
//	<root_hash base64std_padded>\n
//
// The trailing \n on the root hash line is part of the authenticated content.
// Stripping it is the most common implementation bug.
package checkpoint

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"strconv"
	"strings"
)

// Body formats the checkpoint body that is signed by the issuer and cosigned
// by witnesses. The result is the exact byte sequence over which signatures
// are computed.
//
// IMPORTANT: The root hash is base64-encoded with standard alphabet and padding
// (RFC 4648 §4). The exact encoding matters for signature verification.
// Implementations that use URL-safe base64 or strip padding will fail.
func Body(origin string, treeSize uint64, rootHash []byte) []byte {
	rootHashB64 := base64.StdEncoding.EncodeToString(rootHash)
	body := origin + "\n" +
		strconv.FormatUint(treeSize, 10) + "\n" +
		rootHashB64 + "\n"
	return []byte(body)
}

// Sign signs the checkpoint body with an Ed25519 private key.
// Returns the raw 64-byte signature.
func Sign(body []byte, key ed25519.PrivateKey) []byte {
	return ed25519.Sign(key, body)
}

// Verify verifies an Ed25519 signature over a checkpoint body.
func Verify(body []byte, sig []byte, key ed25519.PublicKey) error {
	if !ed25519.Verify(key, body, sig) {
		return fmt.Errorf("checkpoint: Ed25519 signature verification failed")
	}
	return nil
}

// KeyID derives the 4-byte key ID for an Ed25519 signing key.
// Per c2sp.org/signed-note:
//
//	key_id = SHA-256(key_name || 0x0A || 0x01 || raw_ed25519_pubkey)[0:4]
//
// where 0x0A is a newline and 0x01 is the Ed25519 signature type identifier byte.
func KeyID(humanName string, pubKey ed25519.PublicKey) [4]byte {
	h := sha256.New()
	h.Write([]byte(humanName))
	h.Write([]byte{0x0A, 0x01}) // newline + Ed25519 type byte
	h.Write(pubKey)
	sum := h.Sum(nil)
	var id [4]byte
	copy(id[:], sum[:4])
	return id
}

// OriginID returns the first 8 bytes of SHA-256(origin), stored big-endian,
// used as the routing hint in the MTA-QR payload.
func OriginID(origin string) uint64 {
	h := sha256.Sum256([]byte(origin))
	var id uint64
	for i := 0; i < 8; i++ {
		id = (id << 8) | uint64(h[i])
	}
	return id
}

// CosignatureV1Message constructs the signed message for a tlog-cosignature/v1.
//
//	cosignature/v1\n
//	time <unix_timestamp_decimal>\n
//	<checkpoint body>
func CosignatureV1Message(body []byte, unixTimestamp uint64) []byte {
	header := "cosignature/v1\n" +
		"time " + strconv.FormatUint(unixTimestamp, 10) + "\n"
	msg := make([]byte, len(header)+len(body))
	copy(msg, header)
	copy(msg[len(header):], body)
	return msg
}

// SignCosignature produces a tlog-cosignature/v1 signature (64 bytes) over
// the cosignature message constructed from body and timestamp.
func SignCosignature(body []byte, timestamp uint64, key ed25519.PrivateKey) []byte {
	msg := CosignatureV1Message(body, timestamp)
	return ed25519.Sign(key, msg)
}

// VerifyCosignature verifies a tlog-cosignature/v1 signature.
// The timestamp in the signed message and in the binary WitnessCosig struct
// must be identical; that identity is the caller's responsibility.
func VerifyCosignature(body []byte, timestamp uint64, sig []byte, key ed25519.PublicKey) error {
	msg := CosignatureV1Message(body, timestamp)
	if !ed25519.Verify(key, msg, sig) {
		return fmt.Errorf("checkpoint: witness cosignature verification failed")
	}
	return nil
}

// ParseBody parses a checkpoint body and returns origin, treeSize, rootHash.
// Returns an error if the body does not have exactly three newline-terminated lines.
func ParseBody(body []byte) (origin string, treeSize uint64, rootHash []byte, err error) {
	s := string(body)
	if !strings.HasSuffix(s, "\n") {
		return "", 0, nil, fmt.Errorf("checkpoint: body must end with newline")
	}
	// Per c2sp.org/tlog-checkpoint: the body is three mandatory lines followed by
	// optional extension lines. We parse the first three and ignore any extras.
	lines := strings.Split(strings.TrimSuffix(s, "\n"), "\n")
	if len(lines) < 3 {
		return "", 0, nil, fmt.Errorf("checkpoint: expected at least 3 lines, got %d", len(lines))
	}
	origin = lines[0]
	treeSize, err = strconv.ParseUint(lines[1], 10, 64)
	if err != nil {
		return "", 0, nil, fmt.Errorf("checkpoint: invalid tree_size %q: %w", lines[1], err)
	}
	rootHash, err = base64.StdEncoding.DecodeString(lines[2])
	if err != nil {
		return "", 0, nil, fmt.Errorf("checkpoint: invalid root_hash base64 %q: %w", lines[2], err)
	}
	if len(rootHash) != 32 {
		return "", 0, nil, fmt.Errorf("checkpoint: root_hash must be 32 bytes, got %d", len(rootHash))
	}
	return origin, treeSize, rootHash, nil
}
